geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java

```diff
@@ -2905,45 +2904,44 @@ public void validatePoolFactory(PoolFactory poolFactory) {
-          }
-          // check for case where a root region is being reinitialized
-          // and we didn't find a region, i.e. the new region is about to be created
 
-          if (!isReInitCreate) {
-            String fullPath = Region.SEPARATOR + name;
-            future = reinitializingRegions.get(fullPath);
-          }
-          if (future == null) {
-            if (internalRegionArgs.getInternalMetaRegion() != null) {
-              region = internalRegionArgs.getInternalMetaRegion();
-            } else if (isPartitionedRegion) {
-              region = new PartitionedRegion(name, attrs, null, this, internalRegionArgs,
-                  statisticsClock, ColocationLoggerFactory.create());
+        region = rootRegions.get(name);
+        if (region != null) {
+          throw new RegionExistsException(region);
+        }
+        // check for case where a root region is being reinitialized
+        // and we didn't find a region, i.e. the new region is about to be created
+
+        if (!isReInitCreate) {
+          String fullPath = Region.SEPARATOR + name;
+          future = reinitializingRegions.get(fullPath);
+        }
+        if (future == null) {
+          if (internalRegionArgs.getInternalMetaRegion() != null) {
+            region = internalRegionArgs.getInternalMetaRegion();
+          } else if (isPartitionedRegion) {
+            region = new PartitionedRegion(name, attrs, null, this, internalRegionArgs,
+                statisticsClock, ColocationLoggerFactory.create());
+          } else {
+            // Abstract region depends on the default pool existing so lazily initialize it
+            // if necessary.
+            if (Objects.equals(attrs.getPoolName(), DEFAULT_POOL_NAME)) {
+              determineDefaultPool();
+            }
+            if (attrs.getScope().isLocal()) {
+              region =
+                  new LocalRegion(name, attrs, null, this, internalRegionArgs, statisticsClock);
             } else {
-              // Abstract region depends on the default pool existing so lazily initialize it
-              // if necessary.
-              if (Objects.equals(attrs.getPoolName(), DEFAULT_POOL_NAME)) {
-                determineDefaultPool();
-              }
-              if (attrs.getScope().isLocal()) {
-                region =
-                    new LocalRegion(name, attrs, null, this, internalRegionArgs, statisticsClock);
-              } else {
-                region = new DistributedRegion(name, attrs, null, this, internalRegionArgs,
-                    statisticsClock);
-              }
+              region = new DistributedRegion(name, attrs, null, this, internalRegionArgs,
+                  statisticsClock);
             }
+          }
 
-            rootRegions.put(name, region);
-            if (isReInitCreate) {
-              regionReinitialized(region);
-            }
-            break;
+          rootRegions.put(name, region);
```

Comment:
This does not look thread safe. I don't think we want to do a get followed by a put here. This allows two threads to both see null on the get and then they both create it and the last one wins. Maybe a single call of computeIfAbsent instead of the get+put is what you want here